nsxt/resource_nsxt_policy_predefined_gateway_policy.go

```diff
@@ -0,0 +1,493 @@
+
+	targetType := "Domain"
+	childDomain := model.ChildResourceReference{
+		Id:           &domain,
+		ResourceType: "ChildResourceReference",
+		TargetType:   &targetType,
+		Children:     domainChildren,
+	}
+
+	dataValue, errors = converter.ConvertToVapi(childDomain, model.ChildResourceReferenceBindingType())
+	if len(errors) > 0 {
+		return nil, errors[0]
+	}
+	return dataValue.(*data.StructValue), nil
+}
+
+func gatewayPolicyInfraPatch(policy model.GatewayPolicy, domain string, m interface{}) error {
+	childDomain, err := createChildDomainWithGatewayPolicy(domain, *policy.Id, policy)
+	if err != nil {
+		return fmt.Errorf("Failed to create H-API for Predefined Gateway Policy: %s", err)
+	}
+
+	var infraChildren []*data.StructValue
+	infraChildren = append(infraChildren, childDomain)
+
+	infraType := "Infra"
+	infraObj := model.Infra{
+		Children:     infraChildren,
+		ResourceType: &infraType,
+	}
+
+	return policyInfraPatch(infraObj, isPolicyGlobalManager(m), getPolicyConnector(m), false)
+
+}
+
+func updatePolicyPredefinedGatewayPolicy(id string, d *schema.ResourceData, m interface{}) error {
+
+	connector := getPolicyConnector(m)
+	isGlobalManager := isPolicyGlobalManager(m)
+	path := d.Get("path").(string)
+	domain := getDomainFromResourcePath(path)
+
+	if domain == "" {
+		return fmt.Errorf("Failed to extract domain from Gateway Policy path %s", path)
+	}
+
+	predefinedPolicy, err := getGatewayPolicyInDomain(id, domain, connector, isPolicyGlobalManager(m))
+	if err != nil {
+		return err
+	}
+
+	if d.HasChange("description") {
+		description := d.Get("description").(string)
+		predefinedPolicy.Description = &description
+	}
+
+	if d.HasChange("tag") {
+		predefinedPolicy.Tags = getPolicyTagsFromSchema(d)
+	}
+
+	var childRules []*data.StructValue
+	if d.HasChange("rule") {
+		oldRules, _ := d.GetChange("rule")
+		rules := getPolicyRulesFromSchema(d, true)
+
+		existingRules := make(map[string]bool)
+		for _, rule := range rules {
+			ruleID := newUUID()
+			if rule.Id != nil {
+				ruleID = *rule.Id
+				existingRules[ruleID] = true
+			} else {
+				rule.Id = &ruleID
+			}
+
+			childRule, err := createPolicyChildRule(ruleID, rule, false)
+			if err != nil {
+				return err
+			}
+			log.Printf("[DEBUG]: Adding child rule with id %s", ruleID)
+			childRules = append(childRules, childRule)
+		}
+
+		for _, oldRule := range oldRules.([]interface{}) {
+			oldRuleMap := oldRule.(map[string]interface{})
+			oldRuleID := oldRuleMap["nsx_id"].(string)
+			if _, exists := existingRules[oldRuleID]; !exists {
+				resourceType := "Rule"
+				rule := model.Rule{
+					Id:           &oldRuleID,
+					ResourceType: &resourceType,
+				}
+
+				childRule, err := createPolicyChildRule(oldRuleID, rule, true)
+				if err != nil {
+					return err
+				}
+				log.Printf("[DEBUG]: Deleting child rule with id %s", oldRuleID)
+				childRules = append(childRules, childRule)
+
+			}
+		}
+	}
+
+	if d.HasChange("default_rule") {
```

Comment:
Will it hit both HasChange("rule") and HasChange("default_rule")? If not shall we use "else if" here?

Reply:
Done